### PR TITLE
Emulated etcd version

### PR DESCRIPTION
**Pull request by louiseschmidtgen**
## Emulated etcd version

Support for etcd version check that evaluates whether RequestWatchProgress is supported by the current version of etcd endpoint from status request: https://github.com/kubernetes/kubernetes/blob/beb696c2c9467dbc44cbaf35c5a4a3daf0321db3/staging/src/k8s.io/apiserver/pkg/storage/feature/feature_support_checker.go#L157

Affected upstream k8s features: `ConsistentListFromCache`, `WatchList`.

Backport of https://github.com/k3s-io/kine/pull/316.

This PR defines an emulated etcd version that signifies the RequestWatchProgress is not yet supported. Only once k8s-dqlite supports WatchProgressRequests we can bump the version to signal the feature's support.

Tested with 1.32 rc: https://github.com/canonical/k8s-dqlite/actions/runs/12237343236

### CODE REVIEW

File: .github/workflows/k8s-snap-integration.yaml [hunk 1]
```diff
@@ -14,7 +14,11 @@ concurrency:
 
 jobs:
   build:
-    name: K8s-snap Integration Test
+    name: K8s-snap Integration Test ${{ matrix.os }}
```
Comment (bschimke95) [inline .github/workflows/k8s-snap-integration.yaml:17]: This looks unrelated

**Review by bschimke95 — COMMENTED** [review]
Mostly LGTM, minor nit

File: pkg/kine/server/get.go [hunk 2]
```diff
@@ -23,11 +23,9 @@ func (l *LimitedServer) get(ctx context.Context, r *etcdserverpb.RangeRequest) (
 		attribute.Int64("revision", r.Revision),
 	)
 
-	if len(r.RangeEnd) != 0 {
-		return nil, fmt.Errorf("unexpected rangeEnd: want empty, got %s", r.RangeEnd)
-	}
-	if r.Limit != 0 {
-		return nil, fmt.Errorf("unexpected limit: want 0, got %d", r.Limit)
+	if r.Limit != 0 && len(r.RangeEnd) != 0 {
```
Comment (berkayoz) [inline pkg/kine/server/get.go:26]: Can you give me more context on this? I'm not sure why we changed this.

**Review by berkayoz — COMMENTED** [review]
LGTM overall, small questions

File: cmd/root.go [hunk 3]
```diff
@@ -173,6 +175,7 @@ func init() {
 	rootCmd.Flags().BoolVar(&rootCmdOpts.otel, "otel", false, "enable traces endpoint")
 	rootCmd.Flags().StringVar(&rootCmdOpts.otelAddress, "otel-listen", "127.0.0.1:4317", "listen address for OpenTelemetry endpoint")
 	rootCmd.Flags().StringVar(&rootCmdOpts.metricsAddress, "metrics-listen", "127.0.0.1:9042", "listen address for metrics endpoint")
+	rootCmd.Flags().StringVar(&rootCmdOpts.emulatedEtcdVersion, "emulated-etcd-version", "3.5.12", "The emulated etcd version to return on a call to the status endpoint. Defaults to 3.5.12, in order to indicate no support for watch progress notifications yet.")
```
[the remark this answers is not in the document]
Comment (bschimke95) [reply cmd/root.go:178]: I think it is fine to expose that as a CLI argument.
Upstream kine does this as well.

File: docs/configuration.md [hunk 4]
```diff
@@ -21,6 +21,7 @@ The following configuration options are available listed in a table format:
 | `--otel` | Enable traces endpoint | `false` |
 | `--otel-listen` | The address to listen for OpenTelemetry endpoint | `127.0.0.1:4317` |
 | `--metrics-listen` | The address to listen for metrics endpoint | `127.0.0.1:9042` |
+| `--emulated-etcd-version` | The emulated etcd version to return on a call to the status endpoint. Defaults to 3.5.7, in order to indicate no support for watch progress notifications yet. | `3.5.7` |
```
[the remark this answers is not in the document]
Comment (louiseschmidtgen) [reply docs/configuration.md]: Thank you, I've updated this :D

File: pkg/kine/server/get.go [hunk 5]
```diff
@@ -23,11 +23,9 @@ func (l *LimitedServer) get(ctx context.Context, r *etcdserverpb.RangeRequest) (
...
+		err := fmt.Errorf("invalid combination of rangeEnd and limit, limit should be 0 got %d", r.Limit)
```
Comment (HomayoonAlimohammadi) [inline pkg/kine/server/get.go:27]: Maybe we can change this to a named return (probably in another PR). Because IMO it's easy to just `return fmt.Errorf(...)` which won't be caught in the defer.

**Review by HomayoonAlimohammadi — APPROVED** [review]
Thanks a lot @louiseschmidtgen, LGTM!

**Comment by louiseschmidtgen** [comment]
See https://github.com/canonical/k8s-dqlite/pull/212/files#r1879804010, emulatedEtcdVersion is implemented in the linked PR.
